htlcswitch/mailbox.go

```diff
@@ -402,3 +402,174 @@ func (m *memoryMailBox) MessageOutBox() chan lnwire.Message {
+// mailbox and removed (marked as claimed).
+func (mo *mailOrchestrator) BindLiveShortChanID(mailbox MailBox,
+	cid lnwire.ChannelID, sid lnwire.ShortChannelID) {
+
+	mo.mu.Lock()
+	// Update the mapping from short channel id to mailbox's channel id.
+	mo.liveIndex[sid] = cid
+
+	// Retrieve any unclaimed packets destined for this mailbox.
+	pkts, _ := mo.unclaimedPackets[sid]
+	delete(mo.unclaimedPackets, sid)
+	mo.mu.Unlock()
+
+	// Deliver the unclaimed packets.
+	for _, pkt := range pkts {
+		mailbox.AddPacket(pkt)
+	}
+}
+
+// Deliver lookups the target mailbox using the live index from short_chan_id
+// to channel_id. If the mailbox is found, the message is delivered directly.
+// Otherwise the packet is recorded as unclaimed, and will be delivered to the
+// mailbox upon the subsequent call to BindLiveShortChanID.
+func (mo *mailOrchestrator) Deliver(
+	sid lnwire.ShortChannelID, pkt *htlcPacket) error {
+
+	var (
+		mailbox MailBox
+		found   bool
+	)
+
+	// First, try to find the channel id for the target short_chan_id. If
+	// the link is live, we will also look up the created mailbox.
+	mo.mu.RLock()
+	chanID, isLive := mo.liveIndex[sid]
+	if isLive {
+		mailbox, found = mo.mailboxes[chanID]
+	}
+	mo.mu.RUnlock()
+
+	// The link is live and target mailbox was found, deliver immediately.
+	if isLive && found {
+		return mailbox.AddPacket(pkt)
+	}
+
+	// If we detected that the link has not been made live, we will acquire
+	// the exclusive lock preemptively in order to queue this packet in the
+	// list of unclaimed packets.
+	mo.mu.Lock()
+
+	// Double check to see if the mailbox has been not made live since the
+	// release of the shared lock.
+	//
+	// NOTE: Checking again with the exclusive lock held prevents a race
+	// condition where BindLiveShortChanID is interleaved between the
+	// release of the shared lock, and acquiring the exclusive lock. The
+	// result would be stuck packets, as they wouldn't be redelivered until
+	// the next call to BindLiveShortChanID, which is expected to occur
+	// infrequently.
+	chanID, isLive = mo.liveIndex[sid]
+	if isLive {
+		// Reaching this point indicates the mailbox is actually live.
+		// We'll try to load the mailbox using the fresh channel id.
+		//
+		// NOTE: This should never create a new mailbox, as the live
+		// index should only be set if the mailbox had been initialized
+		// beforehand.  However, this does ensure that this case is
+		// handled properly in the event that it could happen.
+		mailbox = mo.exclusiveGetOrCreateMailBox(chanID)
+		mo.mu.Unlock()
+
+		// Deliver the packet to the mailbox if it was found or created.
+		return mailbox.AddPacket(pkt)
+	}
+
+	// Finally, if the channel id is still not found in the live index,
+	// we'll add this to the list of unclaimed packets. These will be
+	// delivered upon the next call to BindLiveShortChanID.
+	mo.unclaimedPackets[sid] = append(mo.unclaimedPackets[sid], pkt)
```

Comment:
Hmm, so before this commit, any HTLC that targeted a link which we had a short chan ID for would all go to the _same_ mailbox and never actually be deleted?

Reply:
So we delay queueing these packets into a mailbox until that mapping is finalized